### ISSUE TITLE
erlang-mode is broken with GNU Emacs HEAD

### ISSUE THREAD

**Issue by wahjava** [erlang/otp#5314]
**Describe the bug**

Since emacs-mirror/emacs@86da812afb2572c7fead2bb07570b976bffd7c55 (guessing), `erlang-mode` is broken. Following is a helpful backtrace:

```
Debugger entered--Lisp error: (error "Given parent class xref-file-location is not a cla...")
  signal(error ("Given parent class xref-file-location is not a cla..."))
  error("Given parent class %S is not a class" xref-file-location)
  eieio-defclass-internal(erlang-xref-location (xref-file-location) ((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) (:documentation "An erlang location is a file location plus arity."))
  (progn (defalias 'erlang-xref-location-p (eieio-make-class-predicate 'erlang-xref-location)) (defalias 'erlang-xref-location--eieio-childp (eieio-make-child-predicate 'erlang-xref-location)) (defalias 'erlang-xref-location-child-p #'erlang-xref-location--eieio-childp) (make-obsolete 'erlang-xref-location-child-p "use (cl-typep ... \\='erlang-xref-location) instead" "25.1") (define-symbol-prop 'erlang-xref-location 'cl-deftype-satisfies #'erlang-xref-location--eieio-childp) (eieio-defclass-internal 'erlang-xref-location '(xref-file-location) '((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) '(:documentation "An erlang location is a file location plus arity.")) (cl-defmethod erlang-xref-location-arity ((this erlang-xref-location)) "Access the slot `arity' from object of class `erla..." (slot-value this 'arity)) (defun erlang-xref-location (&rest slots) "Create a new object of class type `erlang-xref-loc..." (declare (compiler-macro (lambda (whole) (if (not (stringp ...)) whole (macroexp-warn-and-return (format "Obsolete name arg %S to constructor %S" ... ...) `...))))) (apply #'make-instance 'erlang-xref-location slots)))
  (defclass erlang-xref-location (xref-file-location) ((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) :documentation "An erlang location is a file location plus arity.")
  (progn (cl-defmethod xref-backend-identifier-at-point ((_backend (eql erlang-etags))) (if (eq this-command 'xref-find-references) (if (use-region-p) (buffer-substring-no-properties (region-beginning) (region-end)) (thing-at-point 'symbol)) (erlang-id-to-string (erlang-get-identifier-at-point)))) (cl-defmethod xref-backend-definitions ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier)) (cl-defmethod xref-backend-apropos ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier t)) (cl-defmethod xref-backend-identifier-completion-table ((_backend (eql erlang-etags))) (let ((erlang-replace-etags-tags-completion-table t)) (tags-completion-table))) (defclass erlang-xref-location (xref-file-location) ((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) :documentation "An erlang location is a file location plus arity.") (cl-defmethod xref-location-marker ((locus erlang-xref-location)) (cl-call-next-method locus)))
  eval((progn (cl-defmethod xref-backend-identifier-at-point ((_backend (eql erlang-etags))) (if (eq this-command 'xref-find-references) (if (use-region-p) (buffer-substring-no-properties (region-beginning) (region-end)) (thing-at-point 'symbol)) (erlang-id-to-string (erlang-get-identifier-at-point)))) (cl-defmethod xref-backend-definitions ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier)) (cl-defmethod xref-backend-apropos ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier t)) (cl-defmethod xref-backend-identifier-completion-table ((_backend (eql erlang-etags))) (let ((erlang-replace-etags-tags-completion-table t)) (tags-completion-table))) (defclass erlang-xref-location (xref-file-location) ((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) :documentation "An erlang location is a file location plus arity.") (cl-defmethod xref-location-marker ((locus erlang-xref-location)) (cl-call-next-method locus))))
  byte-code("\300\301!\203\34\0\300\302!\203\34\0\300\303!\203\34\0\300\304!\203\34\0\305\306!\210\300\207" [erlang-soft-require xref cl-generic eieio etags eval (progn (cl-defmethod xref-backend-identifier-at-point ((_backend (eql erlang-etags))) (if (eq this-command 'xref-find-references) (if (use-region-p) (buffer-substring-no-properties (region-beginning) (region-end)) (thing-at-point 'symbol)) (erlang-id-to-string (erlang-get-identifier-at-point)))) (cl-defmethod xref-backend-definitions ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier)) (cl-defmethod xref-backend-apropos ((_backend (eql erlang-etags)) identifier) (erlang-xref-find-definitions identifier t)) (cl-defmethod xref-backend-identifier-completion-table ((_backend (eql erlang-etags))) (let ((erlang-replace-etags-tags-completion-table t)) (tags-completion-table))) (defclass erlang-xref-location (xref-file-location) ((arity :type fixnum :initarg :arity :reader erlang-xref-location-arity)) :documentation "An erlang location is a file location plus arity.") (cl-defmethod xref-location-marker ((locus erlang-xref-location)) (cl-call-next-method locus)))] 2)
  autoload-do-load((autoload "erlang" "Major mode for editing Erlang source files in Emac..." t nil) erlang-mode)
  command-execute(erlang-mode record)
  execute-extended-command(nil "erlang-mode" "erlang-mode")
  funcall-interactively(execute-extended-command nil "erlang-mode" "erlang-mode")
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```

**To Reproduce**

1. Launch emacs built from current git HEAD (e.g. emacs-mirror/emacs@f35246db1934fa06c96ed8533641a55c899b7a01)

2. In one of the Emacs buffers, do `M-x erlang-mode RET` , assuming `erlang-mode` is in the `load-path`.

**Expected behavior**

`erlang-mode` should get activated.

**Affected versions**
The OTP versions that are affected by this bug.

**Additional context**
Add any other context about the problem here. If you wish to attach Erlang code you can either write it directly in the post using [code tags](https://guides.github.com/features/mastering-markdown/), create a [gist](https://gist.github.com), or attach it as a zip file to this post.

**Comment by juster** [comment]
For the record here are some historical links I had found in my research:

- [Original proposal to change xref internals](https://lists.gnu.org/archive/html/bug-gnu-emacs/2021-09/msg02157.html)
- [Emacs commit which migrates xref from EIEIO to CL-Lib](https://github.com/emacs-mirror/emacs/commit/86da812afb2572c7fead2bb07570b976bffd7c55)

**Comment by rudi** [comment]
This issue also affects released emacs versions: a newer version of the xref package, with the breaking changes, is installed on older emacs versions when the user installs e.g. the eglot or project packages.

I can confirm that @juster's patch fixes this for me on Emacs 27.2.

**Comment by u3s** [comment]
Fix merged in https://github.com/erlang/otp/pull/5324.
Closing the issue.
